packages/core/src/connectors/injected.ts

```diff
@@ -381,21 +380,34 @@ export function injected(parameters: InjectedParameters = {}) {
       const chain = config.chains.find((x) => x.id === chainId)
       if (!chain) throw new SwitchChainError(new ChainNotConfiguredError())
 
+      const chainId_ = numberToHex(chain.id)
+
       try {
         await Promise.all([
-          provider.request({
-            method: 'wallet_switchEthereumChain',
-            params: [{ chainId: numberToHex(chainId) }],
-          }),
-          new Promise<void>((resolve) => {
-            const listener: EIP1193EventMap['chainChanged'] = (data) => {
-              if (Number(data) === chainId) {
-                provider.removeListener('chainChanged', listener)
-                resolve()
+          provider
+            .request({
+              method: 'wallet_switchEthereumChain',
+              params: [{ chainId: chainId_ }],
+            })
+            .then(async () => {
+              // MetaMask makes an `net_version` RPC call to the target chain during `wallet_switchEthereumChain`.
+              // If this request fails, MetaMask doesn't emit a `chainChanged` event, but will still switch the chain.
+              // See https://github.com/MetaMask/metamask-extension/issues/24247
+              //
+              // To work around this, we can ask MetaMask for the current chain ID and emit our internal `changed`
+              // event as if the chain has changed. This will allow the promise below to also resolve. We still want to
+              // wait for the `change` event in case the `eth_chainId` check fails or isn't immediately available.
+              if (
+                chainId_ === (await provider.request({ method: 'eth_chainId' }))
+              ) {
+                config.emitter.emit('change', { chainId })
               }
-            }
-            provider.on('chainChanged', listener)
-          }),
+            }),
+          new Promise<void>((resolve) =>
+            config.emitter.once('change', ({ chainId: currentChainId }) => {
+              if (currentChainId === chainId) resolve()
+            }),
+          ),
```

Comment:
I switched this to use the internal emitter rather than binding to the provider's emitter so that we can emit our own `changed` event in the workaround case. This also matches the MetaMask connector:
https://github.com/wevm/wagmi/blob/72a25ee00a7b6b1b41c1a4825d08440a852f9057/packages/connectors/src/metaMask.ts#L180-L184

Reply:
I believe that either approach can leave this listener open indefinitely if we never get a chain switch event. Should we time out after some time and remove the listener?